File: bindings/java/src/io/quad4/rns/Path.java
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2024-2026 Quad4.io

package io.quad4.rns;

import com.sun.jna.ptr.LongByReference;
import io.quad4.rns.ffi.RnsLibrary;
import java.util.ArrayList;
import java.util.List;

/** Path table helpers. */
public final class Path {
    private Path() {}

    public static final class Info {
        public final byte[] hash;
        public final byte[] via;
        public final int hops;
        public final String iface;
        public final double timestamp;
        public final double expires;

        Info(byte[] hash, byte[] via, int hops, String iface, double timestamp, double expires) {
            this.hash = hash;
            this.via = via;
            this.hops = hops;
            this.iface = iface;
            this.timestamp = timestamp;
            this.expires = expires;
        }
    }

    public static void request(Node node, byte[] destHash) {
        if (destHash == null || destHash.length != Rns.HASH_LEN) {
            throw new RnsException(RnsException.INVALID_ARG);
        }
        Rns.check(RnsLibrary.INSTANCE.rns_path_request(node.handle(), destHash));
    }

    public static List<Info> table(Node node, int capacity, int maxHops) {
        if (capacity <= 0) {
            throw new RnsException(RnsException.INVALID_ARG);
        }
        RnsLibrary.RnsPathEntry[] entries =
                (RnsLibrary.RnsPathEntry[]) new RnsLibrary.RnsPathEntry().toArray(capacity);
        LongByReference written = new LongByReference();
        int code = RnsLibrary.INSTANCE.rns_path_table(node.handle(), entries, capacity, written, maxHops);
        if (code != RnsLibrary.RNS_OK && code != RnsLibrary.RNS_ERR_TRUNCATED) {
            Rns.check(code);
        }
        List<Info> out = new ArrayList<>((int) written.getValue());
        for (int i = 0; i < written.getValue(); i++) {
            RnsLibrary.RnsPathEntry e = entries[i];
            out.add(
                    new Info(
                            RnsLibrary.copyHash(e.hash, e.hash_len),
                            RnsLibrary.copyHash(e.via, e.via_len),
                            e.hops & 0xff,
                            RnsLibrary.cstr(e.iface),
                            e.timestamp,
                            e.expires));
        }
        return out;
    }

    public static boolean known(Node node, byte[] destHash) {
        if (destHash == null || destHash.length != Rns.HASH_LEN) {
            return false;
        }
        try {
            for (Info entry : table(node, 256, -1)) {
                if (Rns.hashEq(entry.hash, destHash)) {
                    return true;
                }
            }
        } catch (RnsException ignored) {
            return false;
        }
        return false;
    }
}

File: bindings/lua/rns/path.lua
-- SPDX-License-Identifier: Apache-2.0
-- Copyright (c) 2024-2026 Quad4.io

local ffi_mod = require("rns.ffi")
local errors = require("rns.errors")
local ffi = ffi_mod.ffi
local lib = ffi_mod.lib

local function path_request(node, dest_hash)
	if type(dest_hash) ~= "string" or #dest_hash ~= ffi_mod.HASH_LEN then
		errors.raise(errors.Error.INVALID_ARG)
	end
	local buf = ffi.new("uint8_t[?]", ffi_mod.HASH_LEN)
	ffi.copy(buf, dest_hash, ffi_mod.HASH_LEN)
	errors.map_code(lib.rns_path_request(node:handle(), buf))
end

local function path_table(node, capacity, max_hops)
	capacity = capacity or 256
	max_hops = max_hops or -1
	if capacity <= 0 then
		errors.raise(errors.Error.INVALID_ARG)
	end
	local raw = ffi.new("rns_path_entry[?]", capacity)
	local written = ffi.new("size_t[1]", 0)
	local code = lib.rns_path_table(node:handle(), raw, capacity, written, max_hops)
	if code ~= errors.Error.OK and code ~= errors.Error.TRUNCATED then
		errors.map_code(code)
	end
	local out = {}
	for i = 0, tonumber(written[0]) - 1 do
		local e = raw[i]
		out[#out + 1] = {
			hash = ffi.string(e.hash, e.hash_len),
			via = ffi.string(e.via, e.via_len),
			hops = tonumber(e.hops),
			iface = ffi_mod.cstr(e.iface, 64),
			timestamp = tonumber(e.timestamp),
			expires = tonumber(e.expires),
		}
	end
	return out
end

local function path_known(node, dest_hash)
	if type(dest_hash) ~= "string" or #dest_hash ~= ffi_mod.HASH_LEN then
		return false
	end
	local ok, table = pcall(path_table, node)
	if not ok then
		return false
	end
	for _, entry in ipairs(table) do
		if entry.hash == dest_hash then
			return true
		end
	end
	return false
end

return {
	path_request = path_request,
	path_table = path_table,
	path_known = path_known,
}
